File: jtaf-spring/src/main/java/ch/jtaf/report/ClubRanking.java
package ch.jtaf.report;

import ch.jtaf.i18n.I18n;
import ch.jtaf.vo.ClubRankingVO;
import ch.jtaf.vo.ClubResultVO;
import com.itextpdf.text.Document;
import com.itextpdf.text.DocumentException;
import com.itextpdf.text.PageSize;
import com.itextpdf.text.pdf.PdfPTable;
import com.itextpdf.text.pdf.PdfWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Locale;

public class ClubRanking extends Ranking {

    private final static Logger LOGGER = LoggerFactory.getLogger(ClubRanking.class);

    private Document document;
    private final ClubRankingVO ranking;

    public ClubRanking(ClubRankingVO ranking, Locale locale) {
        super(locale);
        this.ranking = ranking;
    }

    public byte[] create() {
        try {
            byte[] ba;
            try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
                float border = cmToPixel(1.5f);
                document = new Document(PageSize.A4, border, border, border, border);
                PdfWriter pdfWriter = PdfWriter.getInstance(document, baos);
                pdfWriter.setPageEvent(new HeaderFooter(
                        I18n.getInstance().getString(locale, "Club Ranking"),
                        ranking.getSeries().getName(),
                        ""));
                document.open();
                createRanking();
                document.close();
                pdfWriter.flush();
                ba = baos.toByteArray();
            }

            return ba;
        } catch (DocumentException | IOException e) {
            LOGGER.error(e.getMessage(), e);
            return new byte[0];
        }
    }

    private void createRanking() throws DocumentException {
        PdfPTable table = new PdfPTable(new float[]{2f, 10f, 10f});
        table.setWidthPercentage(100);
        table.setSpacingBefore(cmToPixel(1f));

        int position = 1;
        for (ClubResultVO cr : ranking.getClubs()) {
            createClubRow(table, position, cr);
            position++;
        }
        document.add(table);
    }

    private void createClubRow(PdfPTable table, int position, ClubResultVO cr) {
        addCell(table, position + ".");
        addCell(table, cr.getClub().getName());
        addCellAlignRight(table, "" + cr.getPoints());
    }

}

File: src/main/kotlin/ch/jtaf/control/reporting/report/ClubRanking.kt
package ch.jtaf.control.reporting.report

import ch.jtaf.control.reporting.data.ClubRankingData
import ch.jtaf.control.reporting.data.ClubResultData
import com.itextpdf.text.Document
import com.itextpdf.text.PageSize
import com.itextpdf.text.pdf.PdfPTable
import com.itextpdf.text.pdf.PdfWriter
import org.slf4j.LoggerFactory
import java.io.ByteArrayOutputStream
import java.util.*

class ClubRanking(private val ranking: ClubRankingData, locale: Locale) : Ranking(locale) {

    private val logger = LoggerFactory.getLogger(ClubRanking::class.java)

    private var document: Document? = null

    fun create(): ByteArray {
        try {
            ByteArrayOutputStream().use { baos ->
                val border = cmToPixel(1.5f)
                document = Document(PageSize.A4, border, border, border, border)
                val pdfWriter = PdfWriter.getInstance(document!!, baos)
                pdfWriter.pageEvent = HeaderFooter("Club Ranking", ranking.series.name, "")
                document!!.open()
                createRanking()
                document!!.close()
                pdfWriter.flush()
                return baos.toByteArray()
            }
        } catch (e: Exception) {
            logger.error(e.message, e)
            return ByteArray(0)
        }
    }

    private fun createRanking() {
        val table = PdfPTable(floatArrayOf(2f, 10f, 10f))
        table.widthPercentage = 100f
        table.spacingBefore = cmToPixel(1f)

        var position = 1
        ranking.getSortedClubs().forEach {
            createClubRow(table, position, it)
            position++
        }
        document!!.add(table)
    }

    private fun createClubRow(table: PdfPTable, position: Int, cr: ClubResultData) {
        addCell(table, position.toString() + ".")
        addCell(table, cr.club.name)
        addCellAlignRight(table, "" + cr.points)
    }
}
